engine/common/handler/handler_test.go

```diff
@@ -0,0 +1,117 @@
+	event := "test-event"
+
+	q := queue.NewHeroStore(10, unittest.Logger(), metrics.NewNoopCollector())
+	h := handler.NewAsyncEventHandler(unittest.Logger(), q, 2)
+	eventProcessed := make(chan struct{})
+	h.RegisterProcessor(func(originId flow.Identifier, event interface{}) {
+		require.Equal(t, originId, originId)
+		require.Equal(t, event, event)
+		close(eventProcessed)
+	})
+
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx, _ := irrecoverable.WithSignaler(cancelCtx)
+	h.Start(ctx)
+
+	unittest.RequireCloseBefore(t, h.Ready(), 100*time.Millisecond, "could not start handler")
+
+	require.NoError(t, h.Submit(originId, event))
+
+	unittest.RequireCloseBefore(t, eventProcessed, 100*time.Millisecond, "event not processed")
+	cancel()
+	unittest.RequireCloseBefore(t, h.Done(), 100*time.Millisecond, "could not stop handler")
+}
+
+// testAsyncEventHandlerWithMultipleConcurrentEvents is a test helper that tests the async event handler with multiple events.
+// It submits multiple events to the handler and checks if each event is processed by the handler exactly once.
+// The second argument is a function that generates a new origin identifier for each event.
+func testAsyncEventHandlerWithMultipleConcurrentEvents(t *testing.T, idFactory func() flow.Identifier) {
+	size := 10
+	workers := uint(5)
+
+	tc := make([]struct {
+		originId flow.Identifier
+		event    interface{}
+	}, 10)
+
+	for i := 0; i < size; i++ {
+		tc[i].originId = idFactory()
+		tc[i].event = fmt.Sprintf("test-event-%d", i)
+	}
+
+	q := queue.NewHeroStore(uint32(size), unittest.Logger(), metrics.NewNoopCollector())
+	h := handler.NewAsyncEventHandler(unittest.Logger(), q, workers)
+
+	processedEvents := unittest.NewProtectedMap[string, struct{}]()
+	allEventsProcessed := sync.WaitGroup{}
+	allEventsProcessed.Add(size)
+	h.RegisterProcessor(func(originId flow.Identifier, event interface{}) {
+		// check if the event is in the test case
+		require.Contains(t, tc, struct {
+			originId flow.Identifier
+			event    interface{}
+		}{
+			originId: originId,
+			event:    event,
+		})
+
+		// check if the event is processed only once
+		require.False(t, processedEvents.Has(event.(string)))
+		processedEvents.Add(event.(string), struct{}{})
+
+		allEventsProcessed.Done()
+	})
+
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx, _ := irrecoverable.WithSignaler(cancelCtx)
+	h.Start(ctx)
+
+	unittest.RequireCloseBefore(t, h.Ready(), 100*time.Millisecond, "could not start handler")
+
+	for i := 0; i < size; i++ {
+		go func(i int) {
+			require.NoError(t, h.Submit(tc[i].originId, tc[i].event))
+		}(i)
+	}
+
+	unittest.RequireReturnsBefore(t, allEventsProcessed.Wait, 10*time.Second, "events not processed")
+	cancel()
+	unittest.RequireCloseBefore(t, h.Done(), 100*time.Millisecond, "could not stop handler")
+}
+
+func TestAsyncEventHandler_MultipleConcurrentEvents_DistinctIdentifier(t *testing.T) {
```

Comment:
Can you add a comment to this test and TestAsyncEventHandler_MultipleConcurrentEvents_IdenticalIdentifier just mentioning the expected behavior for DistinctIdentifier vs IdenticalIdentifier if any .

Reply:
https://github.com/onflow/flow-go/pull/3911/commits/d4e10e36b831a7b598392ab859287a99a5adfc04